pkg/runctl/controller.go

```diff
@@ -265,13 +265,28 @@ func (c *Controller) syncHandler(key string) error {
 		return nil
 	}
 
+	// Finished and no deletion timestamp, no need to process anything further
```

Comment:
It is really odd that we double check the state for 2nd time. See  fast exit part. The root cause should be you `Get real pipelineRun bypassing cache`. Would you tell me why shall we need to fetch the object again from API server?

Reply:
This is legacy coding I don't want to touch in this fix.